### PR TITLE
s4905819 2D U-Net module

### PR DESCRIPTION
**Pull request by yanghan8458**
Hello Shakes,

I'm submitting my 2D U-Net model implementation for segmenting Hip MRI data in a prostate cancer study using PyTorch.

The following files are being submitted:

- **download.py** - Contains the script for downloading and preparing the dataset.
- **dataset.py** - Contains the custom data loader for the image dataset.
- **modules.py** - Contains the U-Net model architecture.
- **train.py** - Contains the training loop for the model and the hyperparameters.
- **predict.py** - Contains the function to run the trained model on new data.
- **README.MD** - Contains the theory behind the U-Net architecture and instructions on how to use the code.

**Details：**

- **Loss function**: Binary cross entropy with Logits. 
- **Optimizer**: Adam. 
- **Evaluation metric**: Dice similarity coefficient
- **Learning rate**: 0.001
- **Num epochs**: 30
- **Batch size**: 128

I'd appreciate your feedback on my submission. If there are any adjustments needed or any other concerns, please let me know.

Kind regards,
HAN YANG
49058194

### CODE REVIEW

**Comment by yanghan8458** [comment]
Dear staff team，
I am very sorry to receive this message. I would like to clarify that when I submitted this pull request, I submitted the original version. After submitting it, I debugged it on my own computer and found some errors. I forgot to update the submitted pull request and did not notice several errors in it. However, the return value image in the readme document is my latest submitted version, which contradicts my previous version. I deeply apologize for this!
The current code is already the latest version，Thank you！

**Comment by yanghan8458** [comment]
Dear staff team，
I am very sorry to receive this feedback. I would like to clarify that when I submitted this pull request, I submitted the original version. After submitting it, I debugged it on my own computer and found some errors. I forgot to update the submitted pull request and did not notice several errors in it. However, the return value image in the readme document is my latest submitted version, which contradicts my previous version. I deeply apologize for this!
The current code is already the latest version，Thank you！

**Comment by yanghan8458** [comment]
Dear employee team:
I'm sorry to receive this feedback. I would like to clarify that when I submitted this pull request, I submitted the original version. After submitting, I debugged on my own computer and found some errors. I forgot to update the submitted pull request and did not notice several errors. In addition, the return value image in the readme document is what I expected, so it contains the correct return result. I deeply apologize for this!
The current code is already the latest version, thank you

**Comment by yanghan8458** [comment]
Dear employee team:
I'm sorry to receive this feedback. I would like to clarify that when I submitted this pull request, I submitted the original version. After submitting, I debugged on my own computer and found some errors. I forgot to update the submitted pull request and did not notice several errors. In addition, the return value image in the readme document is the result of my multiple debugging attempts, so it contains the correct return value image and does not match the code that has not been updated. I deeply apologize for this!
The current code is already the latest version, thank you

**Comment by yanghan8458** [comment]
Dear employee team:
I'm sorry to receive this feedback. I would like to clarify that when I submitted this pull request, I submitted the original version. After submitting, I debugged on my own computer and found some errors. I forgot to update the submitted pull request and did not notice several errors. In addition, the return value image in the readme document is the result of my multiple debugging attempts, so it contains the correct return value image and does not match the code that has not been updated. I deeply apologize for this!
The current code is already the latest version, thank you

**Comment by yanghan8458** [comment]
Dear employee team:
I'm sorry to receive this feedback. I would like to clarify that when I submitted this pull request, I submitted the original version. After submitting, I debugged on my own computer and found some errors. I forgot to update the submitted pull request and did not notice several errors. In addition, the return value image in the readme document is the result of my multiple debugging attempts, so it contains the correct return value image and does not match the code that has not been updated. I deeply apologize for this!
The current code is already the latest version, thank you

**Comment by yanghan8458** [comment]
Dear employee team，
I'm sorry to receive this feedback. I would like to clarify that when I submitted this pull request, I submitted the original version. After submitting, I debugged on my own computer and found some errors. I forgot to update the submitted pull request and did not notice several errors. In addition, the return value image in the readme document is the result of my multiple debugging attempts, so it contains the correct return value image and does not match the code that has not been updated. I deeply apologize for this!
The current code is already the latest version, thank you！

**Comment by yanghan8458** [comment]
Dear employee team，
I'm sorry to receive this feedback. I would like to clarify that when I submitted this pull request, I submitted the original version. After submitting, I debugged on my own computer and found some errors. I forgot to update the submitted pull request and did not notice several errors. In addition, the return value image in the readme document is the result of my multiple debugging attempts, so it contains the correct return value image and does not match the code that has not been updated. I deeply apologize for this!
The current code is already the latest version, thank you！

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Code and report has some inconsistancies.</td><td>-3</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Easy. UNet2D (Easy Difficulty)</td><td>-10</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Train logs/charts/Images needs to be added.</td><td>-2</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). Submitted on time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-15</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Segmentation is not one-hot encoded, readme doesn't show output, some consistencies. -3 additionally.
